pubky/src/native/internal/pkarr.rs

```diff
@@ -1,36 +1,140 @@
+        &self,
+        keypair: &Keypair,
+        host: Option<&str>,
+        strategy: PublishStrategy,
+    ) -> Result<()> {
+        // Resolve the most recent record.
         let existing = self.pkarr.resolve_most_recent(&keypair.public_key()).await;
 
-        let mut signed_packet_builder = SignedPacket::builder();
+        // Determine which host we should be using.
+        let host_str = match (host, existing.as_ref()) {
+            // Host was explicitly provided.
+            (Some(h), _) => h.to_owned(),
 
-        if let Some(ref existing) = existing {
-            for answer in existing.resource_records("_pubky") {
-                if !answer.name.to_string().starts_with("_pubky") {
-                    signed_packet_builder = signed_packet_builder.record(answer.to_owned());
+            // No host was provided; attempt to extract from the existing record.
+            (None, Some(rec)) => {
+                match Self::extract_host_from_record(rec) {
+                    Some(h) => h,
+                    // If there's an existing record but no embedded host, there's nothing to update.
+                    None => return Ok(()),
+                }
+            }
+
+            // No host was provided, and no existing record, so nothing to publish.
+            (None, None) => return Ok(()),
+        };
+
+        // Determine if we should publish based on the given strategy.
+        let should_publish = match strategy {
+            PublishStrategy::Force => true,
+            PublishStrategy::IfOlderThan => {
+                match existing {
+                    Some(ref record) => {
+                        let now_micros = Timestamp::now();
+                        let record_micros = record.timestamp().as_u64();
+                        now_micros - record_micros > Timestamp::from(self.max_record_age_micros)
+                    }
+                    None => true, // If there's no record yet, we publish.
                 }
             }
+        };
+
+        if should_publish {
+            self.publish_homeserver_inner(keypair, &host_str, existing)
+                .await?;
         }
 
-        let svcb = SVCB::new(0, host.try_into()?);
+        Ok(())
+    }
 
+    /// Internal helper that builds and publishes the PKarr record.
+    /// Uses an optionally pre-resolved record to avoid re-resolving.
+    async fn publish_homeserver_inner(
+        &self,
+        keypair: &Keypair,
+        host: &str,
+        existing: Option<SignedPacket>,
+    ) -> Result<()> {
+        let mut builder = SignedPacket::builder();
+        if let Some(ref record) = existing {
+            // Append any records (except those already starting with "_pubky") to our builder.
+            for answer in record.resource_records("_pubky") {
+                if !answer.name.to_string().starts_with("_pubky") {
+                    builder = builder.record(answer.to_owned());
+                }
+            }
+        }
```

Comment:
Suggest renaming the SignedPacket to packet instead of record.

```suggestion
        if let Some(ref packet) = existing {
            // Append any records (except those already starting with "_pubky") to our builder.
            for answer in packet.resource_records("_pubky") {
                if !answer.name.to_string().starts_with("_pubky") {
                    builder = builder.record(answer.to_owned());
                }
            }
        }
```